client/castApp/src/castinit.c

```diff
@@ -200,6 +200,127 @@ static void addReccasterEnvVarsCallFunc(const iocshArgBuf *args)
+    if(self->current != casterStateInit) {
+        errlogSevPrintf(errlogMinor, "addReccasterExcludePattern called after iocInit() when reccaster might already be connected. Not supported\n");
+        epicsMutexUnlock(self->lock);
+        return;
+    }
+
+    epicsMutexMustLock(self->lock);
+    num_valid_args = argc;
+    /* check for duplicates and NULLs within argv */
+    for (i = 0; i < argc; i++) {
+        if (argv[i] == NULL) {
+            errlogSevPrintf(errlogMinor, "Arg is NULL for addReccasterExcludePattern\n");
+            num_valid_args--;
+        }
+        else if (strcmp(argv[i], "") == 0) {
+            errlogSevPrintf(errlogMinor, "Arg is empty for addReccasterExcludePattern\n");
+            argv[i] = NULL;
+            num_valid_args--;
+        }
+        else {
+            for (j = i + 1; j < argc; j++) {
+                if (argv[j] && (strcmp(argv[i], argv[j]) == 0)) {
+                    errlogSevPrintf(errlogMinor, "Duplicate pattern %s\n", argv[i]);
+                    argv[i] = NULL;
+                    num_valid_args--;
+                }
+            }
+        }
+    }
+
+    /* check for duplicates between argv and existing */
+    for (i = 0; i < argc; i++) {
+        for (j = 0; j < self->num_exclude_patterns; j++) {
+            if (argv[i] && strcmp(argv[i], self->exclude_patterns[j]) == 0) {
+                errlogSevPrintf(errlogMinor, "Pattern %s is already in exclude_patterns list\n", argv[i]);
+                argv[i] = NULL;
+                num_valid_args--;
+            }
+        }
+    }
+
+    int num_new_excludes = self->num_exclude_patterns + num_valid_args;
+
+    if (num_valid_args != 0) { // only realloc if new data
+        new_exclude = calloc(num_new_excludes, sizeof(char*)); // alloc bigger
+        if(new_exclude == NULL) {
+            errlogSevPrintf(errlogMajor, "Error in memory allocation of new_exclude from addReccasterExcludePattern\n");
+            epicsMutexUnlock(self->lock);
+            return;
+        }
+
+        /* copy data */
+        for (i = 0; i < self->num_exclude_patterns; i++) {
+            if ((new_exclude[i] = strdup(self->exclude_patterns[i])) == NULL) {
+                errlogSevPrintf(errlogMinor, "strdup error for copying %s to new_exclude[%zu] from addReccasterExcludePattern\n", self->exclude_patterns[i], i);
+                ret = 1;
+                break;
+            }
+        }
+        
+        /* allocate new */
+        if (!ret) {
+            for (count = 0; count < argc; count++) {
+                if (argv[count] != NULL) {
+                    if ((new_exclude[i] = strdup(argv[count])) == NULL) {
+                        errlogSevPrintf(errlogMinor, "strdup error for copying %s to new_exclude[%zu] from addReccasterExcludePattern\n", argv[i], i);
+                        ret = 1;
+                        break;
+                    }
+                    i++;
+                }
+            }
+        }
```

Comment:
I think that a lot of this code could be simplified by using EPICS linked list structures instead of manually managing arrays and copying them over; for example, the initial function call could just be:
```
for (int i = 0; i < argc; i++) {
  if (argv[i] && argv[i][0]) {
    if (argv[i] not in linked_list) {
      linked_list.append(argv[i]);
    }
  }
}
```
(give or take)

Reply:
Hi Simon, are you referring to `ELLLIST` in `ellLib.h` ?

Reply:
That's the one, yes.